### PR TITLE
moved inview logic into ComponentView

### PR DESCRIPTION
**Pull request by moloko**
#2269 
allowing both the element to be used for inview and the callback function to be specified, so as to allow for plenty of flexibility

### CODE REVIEW

File: src/core/js/views/componentView.js [hunk 1]
```diff
@@ -23,17 +23,58 @@ define([
...
+            switch (visiblePartY) {
+				case 'top':
+					this.hasSeenTop = true;
+					break;
+				case 'bottom':
+					this.hasSeenBottom = true;
+					break;
+				case 'both':
+					this.hasSeenTop = this.hasSeenBottom = true;
+			}
```
Comment (brian-learningpool) [inline src/core/js/views/componentView.js]: Indentation has gone a bit funky here.

Comment (moloko) [reply src/core/js/views/componentView.js]: that's what I get for copy-pasting some of @tomgreenfield 's code ;-)

File: src/core/js/views/componentView.js [hunk 2]
```diff
@@ -6,15 +6,15 @@ define([
...
-            return "component " + 
-            this.model.get('_component') + 
-            "-component " + this.model.get('_id') + 
-            " " + this.model.get('_classes') + 
-            " " + this.setVisibility() +
-            " " + this.setHidden() +
-            " component-" + this.model.get('_layout') + 
-            " nth-child-" + this.model.get("_nthChild") +
-            " " + (this.model.get('_isComplete') ? 'completed' : '');
+            return 'component ' +
```
Comment (oliverfoster) [inline src/core/js/views/componentView.js]:
```suggestion
            return [ 
                'component',
                this.model.get('_component') +'-component',
                this.model.get('_id'),
                this.model.get('_classes'),
                this.setVisibility(),
                this.setHidden(),
                'component-' + this.model.get('_layout'),
                'nth-child-' + this.model.get('_nthChild'),
                (this.model.get('_isComplete') ? 'completed' : '')
            ].join(" ");
```

File: src/core/js/views/componentView.js [hunk 3]
```diff
@@ -6,15 +6,17 @@ define([
...
-            " " + (this.model.get('_isComplete') ? 'completed' : '');
+            return [
+                'component',
+                this.model.get('_component') +'-component',
+                this.model.get('_id'),
+                this.model.get('_classes'),
+                this.setVisibility(),
+                this.setHidden(),
+                'component-' + this.model.get('_layout'),
+                'nth-child-' + this.model.get('_nthChild'),
```
Comment (lc-thomasberger) [inline src/core/js/views/componentView.js]: I think we can start to deprecate the custom nth-child classes in favor to native css selectors.

File: src/core/js/views/componentView.js [hunk 4]
```diff
@@ -23,17 +25,59 @@ define([
...
+         * for an example. Defaults to `view.setCompletionStatus` if not specified.
+         */
+        setupInviewCompletion: function(inviewElementSelector, callback) {
+            this.$inviewElement = this.$(inviewElementSelector || '.component-inner');
+            this.inviewCallback = (callback || this.setCompletionStatus);
+
+            this.$inviewElement.on('inview.componentView', this.onInview.bind(this));
+        },
+
+        removeInviewListener: function() {
```
Comment (NayanKhedkar) [inline src/core/js/views/componentView.js]: I think there might be need to be add some check.
```
removeInviewListener: function() {
      if (!this.$inviewElement) return;
      this.$inviewElement.off('inview.componentView');
 }
```

**Pull request merged**